### PR TITLE
Match selection

### PR DESCRIPTION
**Pull request by YonatanGideoni**
Addresses issue #5 , now once a team is selected you can choose which round you are scoring and in the scores page filter scores based on match if you type a specific team's number.

### CODE REVIEW

File: client/src/components/scores/scores.js [hunk 1]
```diff
@@ -3,23 +3,37 @@
...
-	        </div>
-	    </div>
-	</div>`,
+    <div class="top-bar-left">
+        <ul class="menu">
+            <li>
+                <input name="search" type="text" ng-model="scores.search" placeholder="search Scores" />
+            </li>
+            <li>
+                <select ng-disabled="!scores.teamIsSelected()" ng-model="scores.round">
```
Comment (idanstark42) [inline client/src/components/scores/scores.js]: The entity which this PR is about, is defined in the rules for the past few years as a `match`. That's also its name in the server. Please change all naming to be the same - Match.

File: client/src/components/scores/scores.js [hunk 2]
```diff
@@ -3,23 +3,37 @@
...
-	        </button>
-	        <button class="button" ng-hide="scores.errorScores().length === 0"
-	        ng-click="scores.showErrors = !scores.showErrors">
-	        {{ scores.showErrors ? 'Show all scores' : 'Show only errors'}}
-	        </button>
-	    </div>
-	</div>
-	<div class="top-bar-page">
-	    <div id="scores-list" class="grid-x grid-padding-x small-up-1 medium-up-3 large-up-5">
-	        <div ng-repeat="score in scores.scores()" class="cell grid-y">
```
Comment (idanstark42) [inline client/src/components/scores/scores.js]: Why did you remove this class? It was meant to set equals height to the score cards

File: client/src/components/scoresheet/scoresheet.js [hunk 3]
```diff
@@ -19,6 +19,13 @@ export default {
 					</datalist>
 				</form>
 			</li>
+			<li>
+				<select ng-disabled="scoresheet.teamIsSelected()" ng-model="scoresheet.round">
```
Comment (idanstark42) [inline client/src/components/scoresheet/scoresheet.js]: Is it disabled when the team is selected?

File: client/src/controllers/scores/scores_controller.js [hunk 4]
```diff
@@ -21,12 +21,19 @@ class ScoresController {
 		this.$scope.$on('reload', () => self.load(true))
 		this.Messanger.on('reload', () => self.load(false), true)
 		this.Tournament.teams().then(teams => {
-			this._loading = false
 			this._teamNumberList = []
 			for (var i = 0; i < teams.length; i++) {//Creates list of team numbers.
 				this._teamNumberList.push(teams[i].number)
 			}
 		})
+		this.Tournament.teams_rounds().then(rounds => {
```
Comment (idanstark42) [inline client/src/controllers/scores/scores_controller.js]: `javascript` conventions says that all names must be written in camelCase, so change `teams_rounds` to `teamsRounds`

File: server/dev_router.js [hunk 5]
```diff
@@ -37,5 +48,9 @@ router.use('/table/all', (req, res) => {
   res.json(TABLES)
 })
 
+router.use('/teamsrounds/all', (req, res) => {
```
Comment (idanstark42) [inline server/dev_router.js]: Please look at the issue #17 which this PR is meant to solve.
The way the matches are retrieved from the server is **by team**. This means that you request `/teams/:team_number/matches` and you get a list of the matches **for that team, and that team alone**. This is in order to save the network a lot of data transmitions.
Also notice the JSON format of the response returned to you by the `Tournament` service. This module (the `dev_server`) should be a mock-responder for these requests, and thus should accept request of the same format, and return them in the same format.
This is reflected in the client as well, because it is meant to work with the format described here (single load when the page goes up, and not every team selection), so it won't work in production

File: client/src/components/scores/scores.js (hunk 2 above)
Comment (YonatanGideoni) [reply client/src/components/scores/scores.js]: I did not remove it if I am understanding properly what class you are talking about (the top-bar-left of the div). Perhaps it shows it like this because of indentation/because I used the menu class here and changed a lot of the things inside?

Comment (idanstark42) [reply client/src/components/scores/scores.js]: Look at the diff without spaces (In the "changes" page, under "diff settings" enable "hide whitespace changes"). Then you can see what I mean. the `grid-y` class has been removed from the `cell`.

File: client/src/components/scores/score.js [hunk 6]
```diff
@@ -3,55 +3,55 @@
...
-	<div class="card-divider">
-	    <a editable-text="score.data.referee" buttons="no" blur="submit" onaftersave="score.save()">
-	        {{ score.data.referee || 'No one' }}
-	    </a>
-	    &#160;on table&#160;
-	    <a editable-select="score.data.tableId" buttons="no" blur="submit" onaftersave="score.save()" e-ng-options="table.tableId as table.tableName for table in score.tables">
-	        {{ score.tableText() }}.
-	    </a>
-	</div>
-	<div class="card-section flex-child-shrink">
```
Comment (idanstark42) [inline client/src/components/scores/score.js]: Please bring back the class `flex-child-shrink` which was removed here in this commit

File: client/src/components/scores/score.js [hunk 7]
```diff
@@ -3,55 +3,55 @@
...
+        </div>
+        <div class="alert button" ng-click="score.openDeletionDialog()">
+            <i class="fa" ng-class="score.deleting ? 'fa-circle-notch fa-spin' : 'fa-trash-alt'"></i>
+        </div>
+    </div>
+</div>
+<div class="deletion-modal reveal" data-reveal>
+    <h4>Do you really want to delete this score?</h4>
+    <div class="subheader">This operation is ireversable!</div>
+    <div class="grid-x align-center button-group button-group">
```
Comment (idanstark42) [inline client/src/components/scores/score.js]: There's no need for the `button-group` class to appear twice

File: client/src/components/scores/score.js (hunk 6 above)
Comment (idanstark42) [inline client/src/components/scores/score.js]: Please restore the `flex-child-shrink` class here.

File: client/src/components/scores/score.js (hunk 7 above)
Comment (idanstark42) [inline client/src/components/scores/score.js]: No need to have the `button-group` class here twice

**Comment by idanstark42** [comment]
Please also notice the three comments from the first review that we're marked as "outdated" because you changed the lines they appear on, but they are still relevant. Aspecially the last one, concerning the API and the integration with the `tournament` service

File: client/src/controllers/scores/scores_controller.js [hunk 8]
```diff
@@ -102,6 +109,27 @@ class ScoresController {
...
+	selectedTeamMatches() {
+		let self = this
+		if (this.teamIsSelected()) {
+			if(self._matches){
+				return self._matches
+			}
+			self.Tournament.teamsMatches(parseInt(this.search)).then(matches => {
+				self._matches = matches
+			})
+			return self._matches
```
Comment (idanstark42) [inline client/src/controllers/scores/scores_controller.js]: The point of a promise is to allow async work. If you call it and then immiddiatly try to use its result, it may not be working fast enough (for example, the response didn't return yet from the server) and you would return here an empty array. This function should return a promise. Speaking of which, not all teams would have the same matches. For example there are re-runs, or teams who decided to miss the second round on purpose. Please don't cache the matches

File: client/src/controllers/scores/scores_controller.js [hunk 9]
```diff
@@ -21,12 +21,16 @@ class ScoresController {
...
 			this._teamNumberList = []
 			for (var i = 0; i < teams.length; i++) {//Creates list of team numbers.
 				this._teamNumberList.push(teams[i].number)
 			}
 		})
+
+
+		self.search = ''
+		self.match = null
+		this._loading = false
```
Comment (idanstark42) [inline client/src/controllers/scores/scores_controller.js]: The point of the `_loading` variable is to be `true` until all neccessary data was loaded for this view. In this view, it occurs when the `teams` have been loaded. Please return the `_loading = false` back into the promise

File: client/src/controllers/scores/scores_controller.js [hunk 10]
```diff
@@ -102,6 +109,27 @@ class ScoresController {
 		return badScores;
 	}
 
+	teamIsSelected() {
+		return !this._loading && this._teamNumberList && this._teamNumberList.indexOf(parseInt(this.search)) > -1
```
Comment (idanstark42) [inline client/src/controllers/scores/scores_controller.js]: What is the meaning behind `this._teamNumberList.indexOf(parseInt(this.search)) > -1`? What is the logical connection between the question of wether the team is selected to the search mechanism of the scores. And why is this all in the `scores` and not in the `score`?

File: client/src/services/tournament.js [hunk 11]
```diff
@@ -49,6 +49,23 @@ class Tournament {
...
+		}
+
+		let self = this
+
+		this._teamNumber = teamNumber
+
+		return this.init()
+			.then(() => self.$http.get(`${self.tournament}/teams/${teamNumber}/matches`))
+			.then(response => {
+				self._teamsMatches = response.data
```
Comment (idanstark42) [inline client/src/services/tournament.js]: If you cache it here like the other requests, you have to cache it per `teamNumber`, because different teams may have different matches

File: client/src/controllers/scores/scores_controller.js (hunk 10 above)
Comment (idanstark42) [inline client/src/controllers/scores/scores_controller.js]: What is the meaning behind `this._teamNumberList.indexOf(parseInt(this.search)) > -1`? What is the logical connection between the question of wether the team is selected to the search mechanism of the scores. And why is this all in the `scores` and not in the `score`?

File: client/src/controllers/scores/scores_controller.js (hunk 9 above)
Comment (YonatanGideoni) [reply client/src/controllers/scores/scores_controller.js]: There is no pre-existing promise in this controller. Should I implement one?

File: client/src/controllers/scores/scores_controller.js (hunk 10 above)
Comment (YonatanGideoni) [reply client/src/controllers/scores/scores_controller.js]: This tests if a team number has been input in the search. This is so it is possible to filter by matches only once a team is selected (as was requested in the issue), as different teams can have different matches.

Comment (idanstark42) [reply client/src/controllers/scores/scores_controller.js]: The request was that the match in a score would be editable one the score has a team. The search in this page should search in all fields, with no regards to which exists

File: client/src/controllers/scores/scores_controller.js (hunk 9 above)
Comment (idanstark42) [reply client/src/controllers/scores/scores_controller.js]: The teams loading is a promise, and therefore can be chained with a `then` function

File: client/src/controllers/scoresheet/scoresheet_controller.js [hunk 12]
```diff
@@ -87,13 +97,24 @@ class ScoresheetController {
         if (scoreDiff !== 0 && isFinite(scoreDiff)) {
             this.showScoreDiffAnimation(scoreDiff)
         }
+
+        if (this.match && this.match.indexOf('✔') > -1) {
```
Comment (idanstark42) [inline client/src/controllers/scoresheet/scoresheet_controller.js]: Again, if you use a seperate value in the `select` on the HTML, you could keep the match object or match id, and here just use the `match.complete` property

File: client/src/controllers/scoresheet/scoresheet_controller.js [hunk 13]
```diff
@@ -87,13 +97,24 @@ class ScoresheetController {
         if (scoreDiff !== 0 && isFinite(scoreDiff)) {
             this.showScoreDiffAnimation(scoreDiff)
         }
+
+        if (this.match && this.match.indexOf('✔') > -1) {
+            return '"' + newScore.toString() + '"'
```
Comment (idanstark42) [inline client/src/controllers/scoresheet/scoresheet_controller.js]: 1. Why are we adding quotes around a score for an existing match?
2. You should use string templates here, like this: \`"${newScore.toString()}"\`

File: client/src/controllers/scoresheet/scoresheet_controller.js [hunk 14]
```diff
@@ -138,6 +162,16 @@ class ScoresheetController {
         this.$scope.$broadcast('set default')
     }
 
+    selectedTeamMatches() {
+        let self = this
```
Comment (idanstark42) [inline client/src/controllers/scoresheet/scoresheet_controller.js]: As before, the `self` here is not needed

File: client/src/services/tournament.js [hunk 15]
```diff
@@ -49,6 +50,21 @@ class Tournament {
 			})
 	}
 
+	teamsMatches(teamNumber){
+		if(this._teamsMatches && this._teamsMatches[teamNumber]){
+			return Promise.resolve(this._teamsMatches[teamNumber])
```
Comment (idanstark42) [inline client/src/services/tournament.js]: Very good job with the promises!
Something I just learned today, is that it's better to save the promise, and always return it, instead of returning `Promise.resolve`. LIke this:
```javascript
const _cachedPromise
function promiseCache() {
  if(!_cachedPromise) {
    _chacedPromise = ... // caclulating the promise
  }
  return _cachedPromise
}
```

File: client/src/controllers/scores/scores_controller.js [hunk 16]
```diff
@@ -21,12 +21,16 @@ class ScoresController {
 		this.$scope.$on('reload', () => self.load(true))
 		this.Messanger.on('reload', () => self.load(false), true)
 		this.Tournament.teams().then(teams => {
-			this._loading = false
```
[the remark this answers is not in the document]
Comment (YonatanGideoni) [reply client/src/controllers/scores/scores_controller.js]: Is it not fully resolved only once the code in the .then has finished executing, as only then is the teamNumberList complete?

File: client/src/controllers/scoresheet/scoresheet_controller.js (hunk 13 above)
Comment (YonatanGideoni) [reply client/src/controllers/scoresheet/scoresheet_controller.js]: 1. To signify that this is a previously completed match.

File: client/src/controllers/scores/scores_controller.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (idanstark42) [reply client/src/controllers/scores/scores_controller.js]: Yes. And only then the score has done loading

File: client/src/services/tournament.js (hunk 15 above)
Comment (YonatanGideoni) [reply client/src/services/tournament.js]: This Promise.resolve format is used in all the functions in tournament. Should I just overhaul all of them?
